sdks/go/pkg/beam/registration/registration.tmpl

```diff
@@ -0,0 +1,263 @@
+{{range $funcIn := upto $startFinishBundleMaxIn}}{{range $funcOut := upto 2}}
+func register{{$upperName}}{{$funcIn}}x{{$funcOut}}FuncAndMakeStructWrapper{{(genericTypingRepresentation $funcIn $funcOut true)}}() func(interface{}) reflectx.Func {
+    reflectx.RegisterFunc(reflect.TypeOf((*func({{range $in := upto $funcIn}}{{if $in}}, {{end}}I{{$in}}{{end}}){{if $funcOut}} R0{{end}})(nil)).Elem(), func(fn interface{}) reflectx.Func {
+        return &caller{{$funcIn}}x{{$funcOut}}{{(genericTypingRepresentation $funcIn $funcOut false)}}{fn: fn.(func({{range $in := upto $funcIn}}{{if $in}}, {{end}}I{{$in}}{{end}}){{if $funcOut}} R0{{end}})}
+    })
+
+    return func(fn interface{}) reflectx.Func {
+        return reflectx.MakeFunc(func({{range $in := upto $funcIn}}{{if $in}}, {{end}}a{{$in}} I{{$in}}{{end}}){{if $funcOut}} R0{{end}} {
+            {{if $funcOut}}return {{end}}fn.({{$lowerName}}{{$funcIn}}x{{$funcOut}}{{(genericTypingRepresentation $funcIn $funcOut false)}}).{{$upperName}}({{range $in := upto $funcIn}}{{if $in}}, {{end}}a{{$in}}{{end}})
+        })
+    }
+}
+{{end}}{{end}}
+{{$numParams := $processElementMaxIn}}
+func build{{$upperName}}Wrapper{{(genericTypingRepresentation $numParams 0 true)}}(doFn interface{}) func(interface{}) reflectx.Func {
+    {{$lowerName}}In := -1
+	{{$lowerName}}Out := -1
+	{{$lowerName}}Method := reflect.ValueOf(doFn).MethodByName("{{$upperName}}")
+	if {{$lowerName}}Method.IsValid() {
+		{{$lowerName}}In = {{$lowerName}}Method.Type().NumIn()
+		{{$lowerName}}Out = {{$lowerName}}Method.Type().NumOut()
+	    switch {
+{{range $funcIn := upto $startFinishBundleMaxIn}}
+    case {{$lowerName}}In == {{$funcIn}}:
+            switch { {{range $funcOut := upto 2}}{{$possibleCombos := (possibleBundleLifecycleParameterCombos $funcIn $numParams)}}{{if $possibleCombos}}
+            case {{$lowerName}}Out == {{$funcOut}}:
+{{$first := true}}{{range $funcCombo := $possibleCombos}}{{if $first}}{{$first = false}}                {{else}} else {{end}}if _, ok := doFn.({{$lowerName}}{{$funcIn}}x{{$funcOut}}{{if (or $funcIn $funcOut)}}[{{(join $funcCombo ", ")}}{{if $funcOut}}{{if $funcIn}}, {{end}}error{{end}}]{{end}}); ok {
+                    return register{{$upperName}}{{$funcIn}}x{{$funcOut}}FuncAndMakeStructWrapper{{if (or $funcIn $funcOut)}}[{{(join $funcCombo ", ")}}{{if $funcOut}}{{if $funcIn}}, {{end}}error{{end}}]{{end}}()
+                } {{end}}{{end}}{{end}}
+            default:
+                panic("Invalid signature for {{$upperName}}")
+            }
+{{end}}
+        default:
+            panic("Invalid signature for {{$upperName}}")
+        }
+    }
+    return nil
```

Comment:
If I followed this correctly in the generated file, it looks like it might be possible for a dense innermost if/else ladder to return out, and land at this `return nil`. Is this intentional?

If so, we should just have a comment just above this saying when the nil return is expected to happen, especially since the switch defaults are to panic.

If not, either add a terminal else clause to the inner if/else ladder, or turn it to a switch so it can also have a terminal default.

Reply:
We could get there if either the function doesn't have a StartBundle function (or whatever function we're dealing with here), or if we were unable to infer the types of the StartBundle function.

Returning nil in the first case is definitely the right thing, I updated to panic in the second case. The second should never happen, but I agree its better to be explicit